vortex-array/src/patches.rs

```diff
@@ -220,15 +244,33 @@ impl Patches {
         Ok(Some(Self::new(stop - start, indices, values)))
     }
 
+    // https://docs.google.com/spreadsheets/d/1D9vBZ1QJ6mwcIvV5wIL0hjGgVchcEnAyhvitqWu2ugU
+    const PREFER_MAP_WHEN_PATCHES_OVER_INDICES_LESS_THAN: f64 = 5.0;
+
+    fn is_map_faster_than_search(&self, take_indices: &ArrayData) -> bool {
+        (self.num_patches() as f64 / take_indices.len() as f64)
+            < Self::PREFER_MAP_WHEN_PATCHES_OVER_INDICES_LESS_THAN
+    }
+
     /// Take the indices from the patches.
-    pub fn take(&self, indices: &ArrayData) -> VortexResult<Option<Self>> {
-        if indices.is_empty() {
+    pub fn take(&self, take_indices: &ArrayData) -> VortexResult<Option<Self>> {
+        if self.is_map_faster_than_search(take_indices) {
+            self.take_map(take_indices)
+        } else {
+            self.take_search(take_indices)
+        }
+    }
+
+    pub fn take_search(&self, take_indices: &ArrayData) -> VortexResult<Option<Self>> {
+        let new_length = take_indices.len();
+
+        if take_indices.is_empty() {
             return Ok(None);
         }
 
         // TODO(ngates): plenty of optimisations to be made here
         let take_indices =
-            try_cast(indices, &DType::Primitive(PType::U64, NonNullable))?.into_primitive()?;
+            try_cast(take_indices, &DType::Primitive(PType::U64, NonNullable))?.into_primitive()?;
```

Comment:
maybe we should get rid of this cast? We convert the values to usize anyway for search_sorted so we could directly convert the take_indices to usize

Reply:
Working with integers is not very ergonomic :|.

I think I achieved what you want. I canonicalize the take_indices in `take` and pass a `PrimitiveArray` into `take_search` and `take_map`. Each of those methods use `match_each_integer_ptype` to avoid any casts before the `usize::try_from`. I had to add an `impl From<Infallible> for VortexError` to make that work (e.g. u8 converts to usize without fail).

I also removed all the NativeIndexPType. I think there is a way to make working with indices more ergonomic but it was a half-baked idea. Instead, I inlined all the code inside the macro.